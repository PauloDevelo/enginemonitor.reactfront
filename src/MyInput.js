import React, {Component} from 'react';

import { FormGroup, Label, Input, FormFeedback } from 'reactstrap';

import { FormattedMessage } from 'react-intl';

import PropTypes from 'prop-types';

class MyInput extends Component {
	
	constructor(props){
		super(props);
		
		this.state = {
			isValid: true
		}
		
		this.validate = this.validate.bind(this);
		this.onChangeHandler = this.onChangeHandler.bind(this);
		this.setInputElem = this.setInputElem.bind(this);
	}
	
	validate(){	
		this.setState((prevState, props) => {
			return{
				isValid: this.inputElem.validity.valid,
				errorMessage: this.inputElem.validationMessage
			}
		});
	}
	
	onChangeHandler(event){
		this.validate();
		if(typeof this.props.handleChange === 'function')this.props.handleChange(event);
	}
	
	setInputElem(inputElem){
		this.inputElem = inputElem;
	}
	
	componentDidMount(){
		this.validate();
	}
	
	render(){
		var required = this.props.required !== undefined;
		var min = this.props.min !== undefined?this.props.min:undefined;
		var max = this.props.max !== undefined?this.props.max:undefined;

		return ( 
			<FormGroup className={"form-group"}>
				<Label for={this.props.name}><FormattedMessage {...this.props.label} /></Label>
				<Input 	innerRef={this.setInputElem} 
						name={this.props.name} 
						type={this.props.type} 
						className={"form-control"} 
						value={this.props.value} 
						onChange={this.onChangeHandler} 
						placeholder={this.props.placeholder} 
						required={required} 
						min={min} max={max} 
						invalid={!this.state.isValid}/>
				<FormFeedback>{this.state.errorMessage}</FormFeedback>
			</FormGroup>
		)
	}
}

MyInput.propTypes = {
	name: PropTypes.string,
	min: PropTypes.number,
	max: PropTypes.number,
	label: PropTypes.object.isRequired,
	type: PropTypes.string.isRequired,
	required: PropTypes.bool,
	value: PropTypes.any.isRequired,
	placeholder: PropTypes.string
};

export default MyInput;