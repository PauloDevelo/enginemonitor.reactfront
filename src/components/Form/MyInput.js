import React, { useState, useRef, useEffect } from 'react';
import { FormGroup, Label, Input, FormFeedback } from 'reactstrap';
import { FormattedMessage } from 'react-intl';
import PropTypes from 'prop-types';

export default function MyInput({ label, handleChange, ...props }) {
	const [validity, setValidity] = useState({ isValid: true, errorMessage: undefined });
	const inputElem = useRef();

	const validate = () => setValidity(
		{
			isValid: inputElem.current.validity !== undefined ? inputElem.current.validity.valid : true,
			errorMessage: inputElem.current.validationMessage
		}
	);

	useEffect(() => 
		validate()
	, [inputElem]);
	
	const onChangeHandler = (event) => {
		validate();
		if(typeof handleChange === 'function') handleChange(event);
	};
	
	if (props.type === 'checkbox'){
		return (
			<FormGroup className={"form-group"} check inline={true}>
				<Label check inline="true">
					<FormattedMessage {...label} />{' '}
					<Input ref={inputElem} onChange={onChangeHandler} invalid={!validity.isValid} {...props} />
				</Label>
			</FormGroup>
		);
	}

	return ( 
		<FormGroup className={"form-group"}>
			<Label for={props.name}><FormattedMessage {...label} /></Label>
			<Input 	innerRef={inputElem} 
					className={"form-control"}
					onChange={onChangeHandler} 
					invalid={!validity.isValid}
					{...props}/>
			{!validity.isValid && <FormFeedback>{validity.errorMessage}</FormFeedback>}
		</FormGroup>
	)
}

MyInput.propTypes = {
	name: PropTypes.string.isRequired,
	min: PropTypes.number,
	max: PropTypes.number,
	label: PropTypes.object.isRequired,
	type: PropTypes.string.isRequired,
	required: PropTypes.bool,
	value: PropTypes.any,
	placeholder: PropTypes.string
};