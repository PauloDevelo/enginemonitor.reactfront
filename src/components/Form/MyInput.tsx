import React, { useState, useRef, useEffect } from 'react';
import { FormGroup, Label, Input, FormFeedback } from 'reactstrap';
import ToolTip from '../ToolTip/ToolTip';

import { FormattedMessage } from 'react-intl';

import PropTypes from 'prop-types';

type Props = {
	label: FormattedMessage.MessageDescriptor,
	tooltip?: FormattedMessage.MessageDescriptor,
	handleChange?: (event: any) => void,
	onChanged?: (newValue: any) => void,
	validationTrigger?: number,
	type: any,
	name: string,
	min?: number,
	max?: number,
	required?: boolean,
	readonly?: string,
	children?: JSX.Element[] | JSX.Element
};

export default function MyInput({ label, tooltip, handleChange, onChanged, validationTrigger, children, ...props }: Props) {
	const [validity, setValidity] = useState({ isValid: true, errorMessage: '' });
	const inputElemRef = useRef<any>();

	const validate = () => {
		const inputElem = inputElemRef.current;
		if(inputElem !== undefined){
			setValidity( { isValid: inputElem.validity !== undefined ? inputElem.validity.valid : true, errorMessage: inputElem.validationMessage });
		}
	};

	useEffect(() => {
		if(validationTrigger !== undefined && validationTrigger != 0){
			validate();
		}
	}, [validationTrigger]);

	useEffect(() => {
		setValidity({ isValid: true, errorMessage: '' });
	}, [inputElemRef]);
	
	const onChangeHandler = (event:React.ChangeEvent<HTMLInputElement>):void => {
		validate();
		if(typeof handleChange === 'function') handleChange(event);

		const target:HTMLInputElement = event.target as HTMLInputElement;
		const value = target.type === 'checkbox' ? target.checked : target.value;
		if(typeof onChanged === 'function') onChanged(value);
	};

	let tooltipElement:JSX.Element | undefined = undefined;
	if (tooltip){
		tooltipElement = ToolTip({tooltip});
	}
	
	if (props.type === 'checkbox'){
		const inlineValue = "true";
		return (
			<FormGroup className={"form-group"} check inline={true}>
				<Label check inline={inlineValue as never}>
					<FormattedMessage {...label}/>{tooltipElement}{' '}
					<Input ref={inputElemRef} onChange={onChangeHandler} invalid={!validity.isValid} {...props}/>
				</Label>
			</FormGroup>
		);
	}

	return ( 
		<FormGroup className={"form-group"}>
			<Label for={props.name}><FormattedMessage {...label} />{tooltipElement}</Label>
			<Input 	innerRef={inputElemRef} 
					className={"form-control"}
					onChange={onChangeHandler} 
					invalid={!validity.isValid}
					{...props}>
				{children}
			</Input>
			{!validity.isValid && <FormFeedback>{validity.errorMessage}</FormFeedback>}
		</FormGroup>
	)
}

MyInput.propTypes = {
	onChanged: PropTypes.func,
	validationTrigger: PropTypes.number,
	name: PropTypes.string.isRequired,
	tooltip: PropTypes.object,
	min: PropTypes.number,
	max: PropTypes.number,
	label: PropTypes.object.isRequired,
	type: PropTypes.string.isRequired,
	required: PropTypes.bool,
	value: PropTypes.any,
	placeholder: PropTypes.string,
	children: PropTypes.node,
};