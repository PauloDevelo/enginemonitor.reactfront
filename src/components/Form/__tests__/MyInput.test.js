import ignoredMessages from '../../../testHelpers/MockConsole';
import React from 'react';
import { mount } from 'enzyme';

import MyInput from '../MyInput'

ignoredMessages.push("[React Intl] Could not find required `intl` object. <IntlProvider> needs to exist in the component ancestry. Using default message as fallback.");

const defaultMessage = {
    id: "defaultId",
    defaultMessage: "Default message",
    description: "Label"
};

describe("MyInput", () => {
    it("of type text should render as expected", () => {
        // Arrange

        const wrapper = mount(<MyInput name="name" 	label={defaultMessage} 	    type="text" 	required/>);

        // Assert
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('label').text()).toEqual(defaultMessage.defaultMessage);
        expect(wrapper.find('input').prop('required')).toEqual(true);
        expect(wrapper.find('input').prop('type')).toEqual('text');
        expect(wrapper.find('input').prop('name')).toEqual('name');
    });

    it("of type checkox should render as expected", () => {
        // Arrange

        const wrapper = mount(<MyInput name="name" 	label={defaultMessage} 	    type="checkbox" 	required/>);

        // Assert
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('label').text()).toEqual(defaultMessage.defaultMessage + ' ');
        expect(wrapper.find('input').prop('required')).toEqual(true);
        expect(wrapper.find('input').prop('type')).toEqual('checkbox');
        expect(wrapper.find('input').prop('name')).toEqual('name');
    }); 
});