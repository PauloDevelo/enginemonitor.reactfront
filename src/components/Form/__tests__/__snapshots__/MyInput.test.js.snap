// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyInput of type checkox should render as expected 1`] = `
<Memo()
  label={
    Object {
      "defaultMessage": "Default message",
      "description": "Label",
      "id": "defaultId",
    }
  }
  name="name"
  required={true}
  type="checkbox"
>
  <FormGroup
    check={true}
    className="form-group"
    inline={true}
    tag="div"
  >
    <div
      className="form-group form-check form-check-inline"
    >
      <Label
        check={true}
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <label
          className="form-check-label"
        >
          <FormattedMessage
            defaultMessage="Default message"
            description="Label"
            id="defaultId"
            values={Object {}}
          >
            Default message
          </FormattedMessage>
           
          <Input
            innerRef={
              Object {
                "current": <input
                  class="form-check-input"
                  name="name"
                  required=""
                  type="checkbox"
                />,
              }
            }
            invalid={false}
            name="name"
            onChange={[Function]}
            required={true}
            type="checkbox"
          >
            <input
              className="form-check-input"
              name="name"
              onChange={[Function]}
              required={true}
              type="checkbox"
            />
          </Input>
        </label>
      </Label>
    </div>
  </FormGroup>
</Memo()>
`;

exports[`MyInput of type text should render as expected 1`] = `
<Memo()
  label={
    Object {
      "defaultMessage": "Default message",
      "description": "Label",
      "id": "defaultId",
    }
  }
  name="name"
  required={true}
  type="text"
>
  <FormGroup
    className="form-group"
    tag="div"
  >
    <div
      className="form-group form-group"
    >
      <Label
        for="name"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <label
          className=""
          htmlFor="name"
        >
          <FormattedMessage
            defaultMessage="Default message"
            description="Label"
            id="defaultId"
            values={Object {}}
          >
            Default message
          </FormattedMessage>
        </label>
      </Label>
      <Input
        className="form-control"
        innerRef={
          Object {
            "current": <input
              class="form-control form-control"
              name="name"
              required=""
              type="text"
            />,
          }
        }
        invalid={false}
        name="name"
        onChange={[Function]}
        required={true}
        type="text"
      >
        <input
          className="form-control form-control"
          name="name"
          onChange={[Function]}
          required={true}
          type="text"
        />
      </Input>
    </div>
  </FormGroup>
</Memo()>
`;
